### ISSUE TITLE
Support connection reset on cancellation for generated REST clients

### ISSUE THREAD

**Issue by calebkiage** [quarkusio/quarkus#41971]
### Description

In PR https://github.com/quarkusio/quarkus/pull/41710, cancellation support was added to the reactive REST services. However, the generated REST clients were never updated to support the scenario. What this means is that if the REST clients are used to call reactive endpoints, then the server will never be notified of the cancellation and will keep running the request to completion.

I have a sample project showing this issue attached below.

In the project, calling:
```
curl -X GET --location "http://localhost:9000/call/cancel1"
```
will reset the connection and the service will stop processing, but calling:
```
curl -X GET --location "http://localhost:9000/call/cancel2"
```
will not.

[Sample project](https://github.com/calebkiage/quarkus-kotlin-cancellation-test)

### Implementation ideas

The code below adds cancellation for a Vert.x HttpClient:

```kotlin
withTimeoutOrNull(delayLong) {
    suspendCancellableCoroutine { cont->
        val options = HttpClientOptions().setDefaultHost("localhost").setDefaultPort(9000)
        val client = ctx.vertx().createHttpClient(options)
        client.request(io.vertx.core.http.HttpMethod.GET, "/long/a").onComplete { conn->
            if (conn.succeeded()) {
                cont.invokeOnCancellation {
                    // Trigger a connection reset on cancellation
                    conn.result().reset()
                }
                val request = conn.result()
                request.send().onComplete { resp->
                    if (resp.succeeded()) {
                        val body = resp.result().body()
                        body.onComplete { bodyState->
                            if (bodyState.succeeded()) {
                                cont.resume(bodyState.result().toString(Charsets.UTF_8))
                            } else {
                                cont.resumeWithException(bodyState.cause())
                            }
                        }
                    } else {
                        cont.resumeWithException(resp.cause())
                    }
                }
            } else {
                cont.resumeWithException(conn.cause())
            }
        }
    }
}
```

**Comment by geoand** [comment]
Thanks, I'll have a look next week

**Comment by geoand** [comment]
@cescoffier @vietj what is the proper way to close / cancel an in flight request from the Vertx HTTP Client?

**Comment by cescoffier** [comment]
You cannot cancel a request, you must close the connection. Be very careful with that, as it breaks keep-alive, and if http/2 with streams is used, you close all of them.

**Comment by cescoffier** [comment]
Resetting seems to be at the stream level. Wondering what happens for http/1

**Comment by cescoffier** [comment]
Ok, 

So for http/1 reset closes the connection (and thus may require another connection handshake later).

For http/2 it reset the stream (using the infamous RST_STREAM frame). The server will receive the info and close the stream eventually (which means that it does not guarantee that the server will stop the work immediately or even that you won't receive the answer). The connection is still established and the other streams are not impacted.

**Comment by geoand** [comment]
Thanks for the detailed response!

So it sounds to me like we should probably shy away from reacting to the cancelation of a `Uni` and **not** try to cancel the underlying HTTP client request - so do nothing for this :)

**Comment by calebkiage** [comment]
I'd there a way to find out if a stream has no subscribers? In that situation, I think the right thing to do is close the connection in http/1 and reset the stream in http/2. This should match many other clients including OkHttp and the JDK clients when using:

```java
// future from: client.sendAsync()...
future.cancel(true);
```

**Comment by calebkiage** [comment]
I'd imagine this change would go in the ClientSendRequestHandler's `handle` method:

https://github.com/quarkusio/quarkus/blob/1f69d1d9a16fc5aa4bbc8078fd4458f1fb7f5572/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java#L89-L90

However, I'm unsure of how to listen to the Mutiny reactive stream subscriber disconnecting from the stream...

**Comment by calebkiage** [comment]
Maybe adding a cancel handler runnable to the request context would work? Then when Quarkus hands the user a Uni, it will have configured the cancellation.

Given the following client:

```java
@Path("/long")
@RegisterRestClient
public interface HelloClient {
    @GET
    @Path("a")
    @Produces(MediaType.TEXT_PLAIN)
    Uni<String> longRunning()
}
```

Maybe it's possible to have the returned uni hooked up to run:

```java
// uni is what would have originally been returned when calling longRunning()
uni.onCancellation().invoke {
    // call previously set cancel handler runnable to reset the connection
}
```

**Comment by geoand** [comment]
Yeah, that would be the way to go, but we need to determine whether actually trying to close the connection is the right thing to do.

**Comment by calebkiage** [comment]
Can it be made into an option like `quarkus.rest-client.*.cancellable=true|false`? It can default to false.

**Comment by geoand** [comment]
@calebkiage can you give #41990 a quick shot?
